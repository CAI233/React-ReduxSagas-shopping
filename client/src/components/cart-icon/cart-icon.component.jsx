import React from 'react';
import { CartIconStyledContainer, CartIconSvgStyled, ItemCount } from './cart-icon.styles';

import { connect } from 'react-redux';
import { toggleCartHidden } from '../../redux/cart/cart.actions';

import { selectCartItemsCount } from '../../redux/cart/cart.selectors';

const CartIcon = ({ toggleCartHidden,itemCount }) => {
    return(
        <CartIconStyledContainer onClick={toggleCartHidden} >
            <CartIconSvgStyled />
            <ItemCount className="test" >
                {itemCount}
            </ItemCount>
        </CartIconStyledContainer>
    );
}
// 使用reduce函数来计算购物车商品数量
const mapStateToProps = state => ({
    // select普通使用法(完成笔记)
    itemCount: selectCartItemsCount(state),
});

const mapDispatchToProps = dispatch => ({
    toggleCartHidden: ()=>dispatch( toggleCartHidden() ),
});

export default connect(mapStateToProps,mapDispatchToProps)(CartIcon);