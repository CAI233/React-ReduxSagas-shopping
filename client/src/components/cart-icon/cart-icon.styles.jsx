import styled from 'styled-components';
import { ReactComponent as CartIconSvg } from "../../assets/shopping-bag.svg";

export const CartIconStyledContainer = styled.div`
    width: 45px;
    height: 45px;
    position: relative;
    display: flex;
    align-items: center;
    justify-content: center;
    cursor: pointer;
`;
CartIconStyledContainer.displayName = "CartIconStyledContainer";

export const CartIconSvgStyled = styled(CartIconSvg)`
    width: 24px;
    height: 24px;
`;

CartIconSvgStyled.displayName = "CartIconSvgStyled";

export const ItemCount = styled.span`
    position: absolute;
    font-size: 10px;
    font-weight: bold;
    bottom: 12px;
`;

ItemCount.displayName = "ItemCount";