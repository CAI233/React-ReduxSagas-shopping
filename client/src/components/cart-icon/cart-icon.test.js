/**
 * 实战2: 处理有connect的组件
 */
// 0. 仅导出JSX部分: 用于测试
import { shallow } from 'enzyme';
import React from 'react';
import { CartIcon } from './cart-icon.component';  // 仅导出JSX部分,用于测试

describe( '购物车图标',()=>{
    
    let mockProps,wrapper;
    beforeEach( ()=>{
        mockProps = {
            toggleCartHidden: jest.fn(),
            itemCount: 0
        }
        wrapper = shallow(<CartIcon {...mockProps} />);
    } );

    // 快照备份时，注释使用英文名称防止出现莫名其妙的错误( 完成笔记 )
    it(" Cart-icon Snapshot ",()=>{
        expect(wrapper).toMatchSnapshot();
    });

    it(" 测试单击图标是否促发切换隐藏菜单函数 ",()=>{
        wrapper.find("CartIconStyledContainer").simulate('click');
        expect( mockProps.toggleCartHidden ).toHaveBeenCalled();
    });

    it(" 测试渲染数字是否正确 ",()=>{
        const num = parseInt(wrapper.find("ItemCount").text()); // text()获得组件中嵌套的内容( 完成笔记 )
        expect( num ).toEqual(0);
    });

} );