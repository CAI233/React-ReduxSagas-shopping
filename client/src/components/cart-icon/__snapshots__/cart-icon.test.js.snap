// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`购物车图标  Cart-icon Snapshot  1`] = `ShallowWrapper {}`;

exports[`购物车图标 test 1`] = `ShallowWrapper {}`;

exports[`购物车图标 快照备份 1`] = `undefined`;
